src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Collections/FlatArrayIntervalTree`1.cs

```diff
@@ -0,0 +1,430 @@
+
+            if (!currentTuple.firstTime)
+            {
+                // We're seeing this node for the second time (as we walk back up the left
+                // side of it).  Now see if it matches our test, and if so return it out.
+                if (testInterval(currentNode.Value, start, length, in introspector))
+                {
+                    matches++;
+                    builder.Add(currentNode.Value);
+
+                    if (stopAfterFirst)
+                        return 1;
+                }
+            }
+            else
+            {
+                // First time we're seeing this node.  In order to see the node 'in-order', we push the right side, then
+                // the node again, then the left side.  This time we mark the current node with 'false' to indicate that
+                // it's the second time we're seeing it the next time it comes around.
+
+                if (ShouldExamineRight(array, start, end, currentNodeIndex, in introspector, out var right))
+                    candidates.Push((right, firstTime: true));
+
+                candidates.Push((currentNodeIndex, firstTime: false));
+
+                if (ShouldExamineLeft(array, start, currentNodeIndex, in introspector, out var left))
+                    candidates.Push((left, firstTime: true));
+            }
+        }
+
+        return matches;
+    }
+
+    private static bool ShouldExamineRight<TIntrospector>(
+        SegmentedArray<Node> array,
+        int start,
+        int end,
+        int currentNodeIndex,
+        in TIntrospector introspector,
+        out int rightIndex) where TIntrospector : struct, IIntervalIntrospector<T>
+    {
+        // right children's starts will never be to the left of the parent's start so we should consider right
+        // subtree only if root's start overlaps with interval's End, 
+        if (introspector.GetSpan(array[currentNodeIndex].Value).Start <= end)
+        {
+            rightIndex = GetRightChildIndex(currentNodeIndex);
+            if (rightIndex < array.Length && GetEnd(array[array[rightIndex].MaxEndNodeIndex].Value, in introspector) >= start)
+                return true;
+        }
+
+        rightIndex = 0;
+        return false;
+    }
+
+    private static bool ShouldExamineLeft<TIntrospector>(
+        SegmentedArray<Node> array,
+        int start,
+        int currentNodeIndex,
+        in TIntrospector introspector,
+        out int leftIndex) where TIntrospector : struct, IIntervalIntrospector<T>
+    {
+        // only if left's maxVal overlaps with interval's start, we should consider 
+        // left subtree
+        leftIndex = GetLeftChildIndex(currentNodeIndex);
+        if (leftIndex < array.Length && GetEnd(array[array[leftIndex].MaxEndNodeIndex].Value, in introspector) >= start)
+            return true;
+
+        return false;
+    }
+
+    IEnumerator IEnumerable.GetEnumerator()
+        => GetEnumerator();
+
+    public IEnumerator<T> GetEnumerator()
```

Comment:
>IEnumerator<T> [](http://example.com/codeflow?start=11&length=14)

no desire for a struct based enumerator?

Reply:
preserving existing behavior (which uses yield iterators).  def will look into struct iterator in followup (i'm moving to making this code shared in that pr).

Reply:
In the next pr, this code will be shared with the binary tree impl.